<section><section id="more-efficient-experimental-design" class="titleslide slide level1"><h1>More Efficient Experimental Design</h1></section></section>
<section><section id="overview" class="titleslide slide level1"><h1>Overview</h1></section><section class="slide level2">

<ul>
<li>Parameter optimisation problems</li>
<li>SpearSeq</li>
<li>Recombinant gene expression</li>
</ul>
</section></section>
<section><section id="parameter-optimisation-problems-in-biology" class="titleslide slide level1"><h1>Parameter Optimisation Problems in Biology</h1></section><section id="in-the-lab" class="slide level2">
<h1>In the lab</h1>
<ul>
<li>Codon optimisation</li>
<li>PCR conditions</li>
<li>Protein purification (IPTG etc</li>
<li>Chemical synthesis (reagent ratios/conditions/catalysts)</li>
<li>Optimising brewing</li>
</ul>
</section><section id="computational-problem" class="slide level2">
<h1>Computational problem</h1>
<ul>
<li>Finding the optimal assembly (according to some metric)</li>
<li>Training detection algorithms e.g. motifs, genes etc.</li>
<li>Optimising clustering methods</li>
</ul>
</section><section id="main-example-problems" class="slide level2">
<h1>Main example problems</h1>
<ul>
<li>Given a protein of interest, which sequence will maximise expression? (e.g. <span class="math inline">900<em>n</em><em>t</em> = 300</span> codons <span class="math inline">∴ ≈ 2<sup>300</sup></span> possible sequences)</li>
<li>Given a set of sequencing data, which preprocessing/assembly parameters will produce the most likely assembly?</li>
</ul>
<p>--</p>
</section></section>
<section><section id="section" class="titleslide slide level1"><h1></h1></section></section>
<section><section id="what-is-a-gaussian-process" class="titleslide slide level1"><h1>What is a Gaussian Process?</h1></section></section>
<section><section id="acquisition-functions" class="titleslide slide level1"><h1>Acquisition functions</h1></section></section>
<section><section id="ucb" class="titleslide slide level1"><h1>UCB</h1></section></section>
